oak_functions_service/src/lookup_htbl.rs

```diff
@@ -0,0 +1,575 @@
+// table to be allocated at 2X the size, data is copied from the old table to the new, and the old
+// table is freed.  This results in a 3X expansion in memory usage vs the prior array size.  In
+// contrast, this scheme has only 40% of the table empty and is never resized.
+//
+// This reduce scheme currently limits the number of entries to 2^32.
+
+const CHUNK_BITS: usize = 21;
+const CHUNK_SIZE: usize = 1 << CHUNK_BITS;
+const CHUNK_MASK: usize = CHUNK_SIZE - 1;
+
+#[derive(Default)]
+pub struct LookupHtbl {
+    table: Vec<u8>, // Contains fixed-size entries with index of k/v data in data_chunks.
+    data_chunks: Vec<Box<[u8]>>,
+    max_entries: usize,       // The number at which we must grow the table.
+    allocated_entries: usize, // This is self.table.len() * ENTRY_SIZE.
+    used_data: usize,
+    used_entries: usize,
+}
+
+enum LookupResult {
+    Found(usize),        // (data_index: u40)
+    NotFound(usize, u8), // (table_index: u40, hash_byte: u8)
+}
+
+impl LookupHtbl {
+    // Set the initial size of self.table.  For best speed and memory use, this should be as large
+    // as the number of key/value entries that will be loaded.
+    pub fn reserve(&mut self, max_entries: usize) {
+        assert!(self.used_entries == 0 && self.used_data == 0);
+        // Use a load factor of 60%.
+        let allocated_entries = (5 * max_entries + 1) / 3;
+        self.max_entries = max_entries;
+        self.table = vec![0u8; ENTRY_SIZE * allocated_entries];
+        self.allocated_entries = allocated_entries;
+        // Allocate 1 extra byte, so the 0 index can be used as NULL.
+        self.data_chunks = vec![];
+        self.used_data = 1usize;
+        self.used_entries = 0usize;
+    }
+
+    // Lookup the key/value pair.  If found, return LookupFound(data_index).  Otherwise, return the
+    // table index that contains 0 which can be used during insert, as NotFound(table_index).
+    fn lookup(&self, key: &[u8]) -> LookupResult {
+        if self.allocated_entries == 0 {
+            panic!("Don't call lookup on empty table");
+        }
+        let key_hash: u64 = hash(key);
+        let key_hash_byte = key_hash as u8;
+        let mut table_index = ENTRY_SIZE * reduce(key_hash as u32, self.allocated_entries as u32);
+        let mut data_index: usize;
+        loop {
+            // We try to load only the hash byte to save time.  Once we have a matching hash byte,
+            // the loop ends and we compare keys.
+            let mut entry_hash_byte = self.table[table_index];
+            while entry_hash_byte != key_hash_byte {
+                if entry_hash_byte == 0 {
+                    data_index = read_index(&self.table, table_index + 1);
+                    if data_index == 0usize {
+                        return LookupResult::NotFound(table_index, key_hash_byte);
+                    }
+                }
+                table_index += ENTRY_SIZE;
+                if table_index == self.table.len() {
```

Comment:
This feels like this algorithm could be simplified.

We have two (wrapping) loops over the same table, each time incrementing the index -- could these be folded into one loop? Also, this loop not having a clear termination condition makes me a bit worried.

Reply:
I agree.  I've factored out some of the code in the loop as a sub-function.  This is a critical loop for speed, so I've not simplified it further.  The inner loop looks for matching hash_bytes.  The outer loop compares the keys.  The special cases makes the loop hard to read.  I'm open to suggestions, but I don't want to sacrifice speed here for readability.